### DISCUSSION TITLE
Thread model for Embedding WAMR in a host

### DISCUSSION THREAD

**Discussion by schorg** [bytecodealliance/wasm-micro-runtime#2355] [Q&A] [answered]
I read in the documentation for the thread model of the [wasm-c-api](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/doc/wasm_c_api.md) that a `wasm_engine_t` instance  may only be created once per process, while every `wasm_store_t` and it's objects may only be accessed in a single thread.

Question: Is this the general thread model for embedding WAMR ?

Background: I have a multi-threaded application and want to initialize the runtime once in the process
```
  /* initialize the wasm runtime by default configurations */
  wasm_runtime_init();

  /* add line below if we want to export native functions to WASM app */
  wasm_runtime_register_natives(...);
```

And then load and run modules from different threads
```
  /* read WASM file into a memory buffer */
  buffer = read_wasm_binary_to_buffer(…, &size);

  /* parse the WASM file from buffer and create a WASM module */
  module = wasm_runtime_load(buffer, size, error_buf, sizeof(error_buf));

  /* create an instance of the WASM module (WASM linear memory is ready) */
  module_inst = wasm_runtime_instantiate(module, stack_size, heap_size,
                                         error_buf, sizeof(error_buf));
```
Are the functions to load and instantiate the modules thread-safe?

**Comment by lum1n0us** [comment, marked as answer]
It is a pretty simple thread model for embedding hosts. It requires a host to create and maintain a thread pool and create an instance for every worker in the pool. Only need to load .wasm once. 

If tends to follow the *"wasm-c-api"* model, it is better to only use APIs in [wasm-c-api.h](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/include/wasm_c_api.h)

**Reply by schorg** [reply]
Thanks for the quick answer and your help.
